### PR TITLE
update macos builds in CI to use new m1 machines

### LINKED ISSUE #3902: handle M1 deprecation in circle ci
[CircleCI is deprecating macos intel support on 31 January 2024](https://discuss.circleci.com/t/macos-intel-support-deprecation-in-january-2024/48718). We should move to a different circleci resource.  We must consider the implications on this "flip" from building x86 to M1 and the impact on the end user, and the preference is to keep a single Mac binary.  A constriant/requirement that things have to change from requiring Rosetta on M1 to requiring Rosetta on x86 seems practical, but we should test that early.

### PR DESCRIPTION
**Pull request by garypen**
Update the build pipeline to produce apple silicon binaries for macos. Additionally, update the installer script to reflect the changes in binary architecture for macOS.

Fixes #3902 

also see: #4483 

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

The installer script was tested manually to make sure it did the "right thing" with respect to no longer modifying _cputype.

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]

File: .changesets/maint_garypen_m1_changes.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+### update macOS builds in CI to use new m1 machines ([Issue #3902](https://github.com/apollographql/router/issues/3902))
```
Comment (abernix) [inline .changesets/maint_garypen_m1_changes.md:1]: For this user-facing heading, I think this title should highlight the outcome/impact more than the technical implementation detail (based on title, this sounds like it is just a CI change):

```suggestion
### Pre-built binaries are only available for Apple Silicon ([Issue #3902](https://github.com/apollographql/router/issues/3902))
```

File: .changesets/maint_garypen_m1_changes.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+Prior to this release, macOS binaries were produced on Intel build machines in our CI pipeline. The binaries produced would also work on Apple Silicon (M1, etc.. chips) through the functionality provided by [Rosetta2](https://support.apple.com/en-gb/HT211861).
+
+Our [CI provider has announced the deprecation of the macOS Intel build machines](https://discuss.circleci.com/t/macos-intel-support-deprecation-in-january-2024/48718) and we are updating our build pipeline to use the new Apple Silicon based machines.
+
+This will have the following effects:
+ - Older, Intel based, macOS systems will no longer be able to execute our macOS router binaries
+ - Newer, Apple Silicon based, macOS systems will get a performance boost due to no longer requiring Rosetta2 support.
+
+We have raised an [issue](https://github.com/apollographql/router/issues/4483) that describes options for Intel based macOS users. Please let us know in that issue if the alternatives we suggest (Docker, source build) don't work for you and you'd like us to add support for universal macOS binaries.
```
Comment (abernix) [inline .changesets/maint_garypen_m1_changes.md:11]:
```suggestion
We have raised [an issue](https://github.com/apollographql/router/issues/4483) that describes options for Intel based macOS users. Please let us know in that issue if the alternatives we suggest (e.g., Docker, source build) don't work for you so we can discuss alternatives.
```

File: scripts/install.sh [hunk 3]
```diff
@@ -97,14 +97,15 @@ get_architecture() {
...
-        # but we want to serve x86_64 binaries anyway so that they can
-        # then run in x86_64 emulation mode on their arm64 devices
-        _cputype=x86_64
+    # If our OS is Darwin, we only have binaries for arm64 machines.
+    # Check and exit early
+    if [ "$_ostype" = Darwin ] && [ "$_cputype" != arm64 ]; then
+        say "No precompiled binaries available on mac os for CPU architecture: $_cputype"
+        say "The most recent mac os release which supports intel is 1.37.0"
+        say "Please refer to this issue for more details:  https://github.com/apollographql/router/issues/4483"
+        err "no precompiled binaries available on mac os for CPU architecture: $_cputype"
```
Comment (abernix) [inline scripts/install.sh:106]:
```suggestion
        say "No precompiled binaries available on macOS for CPU architecture: $_cputype"
        say "The most recent macOS release which supports Intel is 1.37.0"
        say "Please refer to this issue for more details:  https://github.com/apollographql/router/issues/4483"
        err "No precompiled binaries available on macOS for CPU architecture: $_cputype"
```

**Review by abernix — CHANGES_REQUESTED** [review]

**Review by abernix — APPROVED** [review]
LGTM!

**Comment by abernix** [comment]
For posterity: See follow-up PRs #4415 and #4418.

**Pull request merged**